vowpalwabbit/cb_explore_adf_rnd.cc

```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+template <bool is_learn>
+void cb_explore_adf_rnd::make_fake_rnd_labels(multi_ex& examples)
+{
+  if (is_learn)
+    {
+      for (size_t i = 0; i < examples.size(); ++i)
+        {
+          auto* ec = examples[i];
+          if (is_the_labeled_example(ec))
+            {
+              ec->l.cb.costs[0].cost = alpha * all->get_random_state()->get_and_update_gaussian() + initials[i];
+              ec->l.cb.costs[0].probability = 1.0f;
+              break;
+            }
+        }
+    }
+}
+
+template <bool is_learn>
+void cb_explore_adf_rnd::restore_labels(multi_ex& examples)
+{
+  if (is_learn)
+    {
+      for (auto* ec: examples)
+        {
+          if (is_the_labeled_example(ec))
+            {
+              ec->l.cb.costs[0].cost = save_class.cost;
+              ec->l.cb.costs[0].probability = save_class.probability;
+              break;
+            }
+        }
+    }
+}
+
+template <bool is_learn>
+void cb_explore_adf_rnd::base_learn_or_predict(VW::LEARNER::multi_learner& base, multi_ex& examples, uint32_t id)
+{
+  if (is_learn)
+    {
+      base.learn(examples, id);
+    }
+  else
+    {
+      base.predict(examples, id);
+    }
+}
+
+template <bool is_learn>
+void cb_explore_adf_rnd::predict_or_learn_impl(VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  save_labels<is_learn>(examples);
+  zero_bonuses(examples);
+  for (uint32_t id = 0; id < numrnd; ++id)
+    {
+      get_initial_predictions(examples, 1 + id);
+      make_fake_rnd_labels<is_learn>(examples);
+      base_learn_or_predict<is_learn>(base, examples, 1 + id);
+      accumulate_bonuses(examples);
+    }
+  finish_bonuses();
+  restore_labels<is_learn>(examples);
+  base_learn_or_predict<is_learn>(base, examples, 0);
+
+  v_array<ACTION_SCORE::action_score>& preds = examples[0]->pred.a_s;
+  float maxbonus = std::max(1e-3f, *std::max_element(bonuses.begin(), bonuses.end()));
+  compute_ci(preds, maxbonus);
+  exploration::generate_softmax(
+    -1.0/maxbonus, begin_scores(preds), end_scores(preds), begin_scores(preds), end_scores(preds));
+  exploration::enforce_minimum_probability(epsilon, true, begin_scores(preds), end_scores(preds));
+}
+
+VW::LEARNER::base_learner* setup(VW::config::options_i& options, vw& all)
+{
+  using config::make_option;
+  bool cb_explore_adf_option = false;
+  float epsilon = 0.;
+  float alpha = 0.;
+  float invlambda = 0.;
+  uint32_t numrnd = 1;
+  config::option_group_definition new_options("Contextual Bandit Exploration with Action Dependent Features");
+  new_options
+      .add(make_option("cb_explore_adf", cb_explore_adf_option)
+               .keep()
+               .help("Online explore-exploit for a contextual bandit problem with multiline action dependent features"))
+      .add(make_option("epsilon", epsilon).keep().allow_override().help("minimum exploration probability"))
+      .add(make_option("rnd", numrnd).keep().help("rnd based exploration"))
```

Comment:
Does `rnd` have special meaning? If not, I would say just expand it. Is it rand or round?

Reply:
In the RL literature RND stands for "Random Network Distillation" ...

Reply:
Okay cool. Definitely put that in the file somewhere